### DISCUSSION TITLE
Contributor documentation - how to spin up local backend instance, run and test

### DISCUSSION THREAD

**Discussion by ryukyi** [opennem/opennem#350] [Q&A]
I would love to contribute if possible. I haven't managed to spin up a local instance or create a dev environment or successfully run tests. I couldn't find clear documentation to setup the dev env.

I follow the documentation: https://developers.opennem.org.au/, updated the .env file and ran:

```bash 
docker compose up -d 
```

Would you be able to help me get setup?

**Comment by kahn-jms** [comment]
Hi @ryukyi, the following are the steps I took to get the backend setup, this is everything I needed to do to be able to successfully run `./scripts/init.py` for version/tag `3.16.11`. Note that I've opted for using S3 storage to export the crawler output.

# Steps to get OpenNEM running locally using tag 3.16.11

1. Clone repo and run `git lfs pull`
2. Create `.env` file as copy of `.env.example`
3. Create an S3 bucket for data storage and add the following to your `.env`:
    ```shell
    # Set S3 bucket path to write to
    S3_DATA_BUCKET_PATH=opennem-data
    ```
    with the name of your bucket (I assume your AWS CLI needs to be configured for this to work)
4. Set the ports in `docker-compose.yml` to match that in `.env` (I just removed the `HOST` and set them to `127.0.0.1`)
5. Run:
    ```bash
    poetry env use python3.11
    poetry install
    poetry shell
    sudo docker-compose up -d
    alembic upgrade head
    ```
6. The station `ADPBA1G` is duplicated in `opennem/data/stations.json` so you need to remove one of the entries (I removed the latter).
7. The file `g01hist.xls` from `opennem/datg/au_cpi.json` no longer exists, it’s now `g01hist.xlsx` which the python library `xlrd` can't handle. I downloaded the file from [the RBA site](https://www.rba.gov.au/statistics/tables/) (look for `Consumer Price Inflation – G1`) and converted it to `.xls`, and then pointed the code to that local version in `opennem/core/parsers/cpi.py`:
    ```python
    AU_CPI_URL = "/path/to/g01hist.xls"
    ...
    # Comment out the original code to download the file
    # r = http.get(AU_CPI_URL)
    # if not r.ok:
    #     raise Exception(f"Problem grabbing CPI source: {r.status_code}")
    # Edit this to use `filename` instead of `file_contents`
    wb = xlrd.open_workbook(filename=AU_CPI_URL)
    ```
    Note the change from `file_contents` to `filename` in `xlrd.open_workbook()`.
    This will eventually need to be changed so the code can download the current `.xlsx` file directly and parse it.
8. The return when first querying the (empty) `crawler_meta` table is wrong during the first run, change the following in [/opennem/core/crawlers/meta.py#L50](https://github.com/opennem/opennem/blob/3.16.11/opennem/core/crawlers/meta.py#L50):
    ```python
    if not spider_meta:
        # return None
        return {}
    ```
9. There are extra arguments given to the `run_XX_crawl()` methods in both `opennem/crawlers/mms.py` and `opennem/crawlers/wemde.py`, for both functions add `**kwargs` to the inputs to accept and ignore them.
10. Run `./scripts/init.py` (it will take a long time since it's pulling a lot of history from nemweb).
11. You can then run `./scripts/run_server.py` to spin up the API. You can downloaded the API specification from https://api.opennem.org.au/docs and import it into Postman to test it out (just make sure to set the `baseUrl` variable to point to the server, e.g. `127.0.0.1:8000`).
12. To have the crawlers continuously pull and update the live values, you need to run `huey_consumer` which will execute the tasks in [opennem/workers/scheduler.py](https://github.com/opennem/opennem/blob/3.16.11/opennem/workers/scheduler.py). To make it work I had to comment out the `crawler_run_wem_per_interval` task in lines [opennem/workers/scheduler.py#L75-L78](https://github.com/opennem/opennem/blob/3.16.11/opennem/workers/scheduler.py#L75-L78) (I think it fails because we removed the WEMDE crawler, and because I don’t have an access token for APVI). Run:
    ```bash
    huey_consumer -w4 -f -k process opennem.workers.scheduler.huey
    ```
    This command comes from the prod AWS stack defined in [infra/stacks/prod/docker-compose.yml#L135-L140](https://github.com/opennem/opennem/blob/3.16.11/infra/stacks/prod/docker-compose.yml#L135-L140).

## Notes

- I think the link in [opennem/crawlers/wemde.py#L154](https://github.com/opennem/opennem/blob/3.16.11/opennem/crawlers/wemde.py#L154) might be wrong?
    - It says Trading Report which looks like [tradingReport](https://data.wa.aemo.com.au/public/market-data/wemde/tradingReport/), but it points to the `current/` subdirectory for [referenceTradingPrice](https://data.wa.aemo.com.au/public/market-data/wemde/referenceTradingPrice/current/).
    - My solution was to just comment out all `WEMDE` crawlers in [opennem/crawl.py](https://github.com/opennem/opennem/blob/3.16.11/opennem/crawl.py#L88) (not really a solution but I don’t personally need WEMDE).
    - I also noticed the `current/` directories in the `WEMDE` are empty which I think causes these errors.
- To speed up crawling I uncommented the test value in [opennem/crawlers/mms.py](https://github.com/opennem/opennem/blob/3.16.11/opennem/crawlers/mms.py#L25) and set it to the year 2023.
- The first run will be _really_ slow since it’s crawling the entire NEM history.
